examples/text/plot_document_clustering.py

```diff
@@ -3,284 +3,423 @@
+#
+# We first benchmark the estimators using a dictionary vectorizer along with an
+# IDF normalization as provided by
+# :class:`~sklearn.feature_extraction.text.TfidfVectorizer`.
 
+from sklearn.feature_extraction.text import TfidfVectorizer
 
-# %%
-# Load some categories from the training set
-# ------------------------------------------
+vectorizer = TfidfVectorizer(
+    max_df=0.5,
+    min_df=5,
+    stop_words="english",
+)
+t0 = time()
+X_tfidf = vectorizer.fit_transform(dataset.data)
 
-categories = [
-    "alt.atheism",
-    "talk.religion.misc",
-    "comp.graphics",
-    "sci.space",
-]
-# Uncomment the following to do the analysis on all the categories
-# categories = None
+print(f"vectorizing done in {time() - t0:.3f} s")
+print(f"n_samples: {X_tfidf.shape[0]}, n_features: {X_tfidf.shape[1]}")
 
-print("Loading 20 newsgroups dataset for categories:")
-print(categories)
+# %%
+# After ignoring terms that appear in more than 50% of the documents (as set by
+# `max_df=0.5`) and terms that are not present in at least 5 documents (set by
+# `min_df=5`), the resulting number of unique terms `n_features` is around
+# 8,000. We can additionally quantify the sparsity of the `X_tfidf` matrix as
+# the fraction of non-zero entries devided by the total number of elements.
 
-dataset = fetch_20newsgroups(
-    subset="all", categories=categories, shuffle=True, random_state=42
-)
+print(f"{X_tfidf.nnz / np.prod(X_tfidf.shape):.3f}")
 
-print("%d documents" % len(dataset.data))
-print("%d categories" % len(dataset.target_names))
+# %%
+# We find that less than 1% of the entries of the `X_tfidf` matrix are non-zero.
+#
+# .. _kmeans_sparse_high_dim:
+#
+# Clustering sparse data with K-Means
+# -----------------------------------
+#
+# As both :class:`~sklearn.cluster.KMeans` and
+# :class:`~sklearn.cluster.MiniBatchKMeans` optimize a non-convex objective
+# function, their clustering is not guaranteed to be optimal for a given random
+# init. Even further, on sparse high-dimensional data such as vectorized text,
+# K-means can initialize centroids on extremely isolated data points. Those data
+# points can stay their own centroids all along.
+#
+# The following code illustrates how the previous phenomenon can sometimes lead
+# to highly imbalanced clusters, depending on the random initialization:
+
+from sklearn.cluster import KMeans
+
+for seed in range(5):
+    kmeans = KMeans(
+        n_clusters=true_k,
+        max_iter=100,
+        n_init=1,
+        random_state=seed,
+    ).fit(X_tfidf)
+    cluster_ids, cluster_sizes = np.unique(kmeans.labels_, return_counts=True)
+    print(f"Number of elements asigned to each cluster: {cluster_sizes}")
 print()
+print(f"Real number of elements in each category: {category_sizes}")
+
+# %%
+# To avoid this problem, one possibility is to increase the number of runs with
+# independent random initiations `n_init`. In such case the clustering with the smallest
+# inertia is chosen.
```

Comment:
```suggestion
# To avoid this problem, one possibility is to increase the number of runs with
# independent random initiations `n_init`. In such case the clustering with the best
# inertia (objective function of k-means) is chosen.
```